input/script_utils_test.go

```diff
@@ -2034,3 +2035,134 @@ func TestSpecificationKeyDerivation(t *testing.T) {
+	require.NoError(t, err)
+	randomPub := randomPriv.PubKey()
+	randomPubBytes := randomPub.SerializeCompressed()
+
+	t.ReportAllocs()
+	t.ResetTimer()
+
+	for i := 0; i < t.N; i++ {
+		for j := 0; j < 1000; j++ {
+			runScriptAllocTest(
+				t, dummyData, randomPubBytes, randomPub,
+			)
+		}
+	}
+}
+
+func TestScriptBuilderAlloc(t *testing.T) {
+	dummyData := []byte("dummy data")
+	randomPriv, err := btcec.NewPrivateKey()
+	require.NoError(t, err)
+	randomPub := randomPriv.PubKey()
+	randomPubBytes := randomPub.SerializeCompressed()
+
+	runScriptAllocTest(
+		t, dummyData, randomPubBytes, randomPub,
+	)
+}
+
+func runScriptAllocTest(t testing.TB, dummyData, randomPubBytes []byte,
+	randomPub *btcec.PublicKey) {
+
+	script, err := WitnessScriptHash(dummyData)
+	require.NoError(t, err)
+	require.Len(t, script, 34)
+
+	script, err = WitnessPubKeyHash(randomPubBytes)
+	require.NoError(t, err)
+	require.Len(t, script, 22)
+
+	script, err = GenerateP2SH(dummyData)
+	require.NoError(t, err)
+	require.Len(t, script, 23)
+
+	script, err = GenerateP2PKH(dummyData)
+	require.NoError(t, err)
+	require.Len(t, script, 25)
+
+	script, err = GenMultiSigScript(
+		randomPubBytes, randomPubBytes,
+	)
+	require.NoError(t, err)
+	require.Len(t, script, 71)
+
+	script, err = SenderHTLCScript(
+		randomPub, randomPub, randomPub, dummyData,
+		false,
+	)
+	require.NoError(t, err)
+	require.Len(t, script, 133)
+
+	script, err = SenderHTLCScript(
+		randomPub, randomPub, randomPub, dummyData,
+		true,
+	)
+	require.NoError(t, err)
+	require.Len(t, script, 136)
+
+	script, err = ReceiverHTLCScript(
+		math.MaxUint32, randomPub, randomPub, randomPub,
+		dummyData, false,
+	)
+	require.NoError(t, err)
+	require.Len(t, script, 141)
+
+	script, err = ReceiverHTLCScript(
+		math.MaxUint32, randomPub, randomPub, randomPub,
+		dummyData, true,
+	)
+	require.NoError(t, err)
+	require.Len(t, script, 144)
+
+	script, err = SecondLevelHtlcScript(
+		randomPub, randomPub, math.MaxUint32,
+	)
+	require.NoError(t, err)
+	require.Len(t, script, 80)
+
+	script, err = LeaseSecondLevelHtlcScript(
+		randomPub, randomPub, math.MaxUint32,
+		math.MaxUint32,
+	)
+	require.NoError(t, err)
+	require.Len(t, script, 88)
+
+	script, err = CommitScriptToSelf(
+		math.MaxUint32, randomPub, randomPub,
+	)
+	require.NoError(t, err)
+	require.Len(t, script, 80)
+
+	script, err = LeaseCommitScriptToSelf(
+		randomPub, randomPub, math.MaxUint32,
+		math.MaxUint32,
+	)
+	require.NoError(t, err)
+	require.Len(t, script, 88)
+
+	script, err = CommitScriptUnencumbered(randomPub)
+	require.NoError(t, err)
+	require.Len(t, script, 22)
+
+	script, err = CommitScriptToRemoteConfirmed(randomPub)
+	require.NoError(t, err)
+	require.Len(t, script, 37)
+
+	script, err = LeaseCommitScriptToRemoteConfirmed(
+		randomPub, math.MaxUint32,
+	)
+	require.NoError(t, err)
+	require.Len(t, script, 45)
+
+	script, err = CommitScriptAnchor(randomPub)
+	require.NoError(t, err)
+	require.Len(t, script, 40)
```

Comment:
Got curious so I ran a version without the `require.Len`, and the results got slightly better with a 52.34% reduction in bytes per operation.
```
goos: darwin
goarch: arm64
pkg: github.com/lightningnetwork/lnd/input
                      │  old.prof   │              new.prof              │
                      │   sec/op    │   sec/op     vs base               │
ScriptBuilderAlloc-10   8.944m ± 2%   8.882m ± 2%  -0.69% (p=0.023 n=10)

                      │   old.prof    │               new.prof               │
                      │     B/op      │     B/op      vs base                │
ScriptBuilderAlloc-10   11.749Mi ± 0%   5.600Mi ± 0%  -52.34% (p=0.000 n=10)

                      │  old.prof   │               new.prof               │
                      │  allocs/op  │  allocs/op    vs base                │
ScriptBuilderAlloc-10   77.00k ± 0%   111.00k ± 0%  +44.16% (p=0.000 n=10)
```

Reply:
Did you just run it like this?
```
	_, err = WitnessScriptHash(dummyData)
	require.NoError(t, err)
```

I think the compiler just optimizes away the return value, since it's not used. Could that explain the further 52% reduction?

Reply:
I think `require.Len` needs to put the value on the stack since it's referenced outside the function. Validated with a dummy test, which showed no alloc,
```go
script, err = CommitScriptAnchor(randomPub)
require.NoError(t, err)
script[0] = 1 // make a dummy usage of the returned value
```

Reply:
Okay, I refactored the test a little bit to not use `require` inside the benchmark. Not sure if this is any better.